src/software/geom/algorithms/step_along_perimeter.cpp

```diff
@@ -0,0 +1,81 @@
+#include "software/geom/algorithms/step_along_perimeter.h"
+
+#include <algorithm>
+#include <vector>
+
+#include "software/geom/algorithms/collinear.h"
+#include "software/geom/algorithms/distance.h"
+#include "software/geom/segment.h"
+
+
+Point stepAlongPerimeter(const Polygon& polygon, const Point& start,
+                         double travel_distance)
+{
+    if (travel_distance == 0.0)
+    {
+        return start;
+    }
+
+    const std::vector<Segment>& polygon_segments = polygon.getSegments();
+    std::size_t start_segment_index              = 0;
+
+    // find initial segment which contains start point
+    auto it =
+        std::find_if(polygon_segments.begin(), polygon_segments.end(),
+                     [&start](const Segment& segment) {
+                         return collinear(segment.getStart(), start, segment.getEnd());
+                     });
+
+    if (it == polygon_segments.end())
+    {
+        throw std::runtime_error("Point not on polygon");
```

Comment:
```suggestion
        throw std::invalid_argument("Start point not on polygon perimeter");
```

Reply:
actually let's not throw an error at all. AI crashing in the middle of the game is _very_ bad.

Use LOG(WARNING) and return the argument point itself

Reply:
I don't like exceptions either but a lot of our geom code is littered with them lol. In this case I actually think it's not too bad since you really are breaking the contract by passing in a point not on the perimeter. Whatever is using this function definitely needs to make sure that point is on the perimeter (by getting the point from the shape itself), and isn't just some point that we calculated or approximated. 

Alternatively, we could wrap the returned point in std::optional

Reply:
I agree with Arun that ideally we don't crash the program. My main worry would be having some floating point error which results in the passed in point to be just slightly off of the edge of the polygon.
`std::optional` could be an alternative, though it makes the job of the caller harder if it can't guarantee that it will get a value.
One alternative could be to calculate the `closestPoint` to the edge of the polygon and use that instead. So even if the point is off of the edge, we use the closest point on the polygon. In fact, I had implemented this method in an old PR that never got merge that you can pull into this PR: https://github.com/UBC-Thunderbots/Software/pull/2813/files#diff-c871a8bcdc98ad35ce2607f7d072014f731fa32dbd28feb62f91d9e2e54ec20b
@williamckha do you think think this would meet the requirements of what you're looking for?

Reply:
the floating point error will probably not be an issue since the `collinear()` method used to check if a point is on a segment uses `almostEqual()` to check for equality. however I agree with Arun as well, crashing the program is probably not a good idea, im not sure if using std::optional is a good idea though.

Reply:
> I agree with Arun that ideally we don't crash the program. My main worry would be having some floating point error which results in the passed in point to be just slightly off of the edge of the polygon. `std::optional` could be an alternative, though it makes the job of the caller harder if it can't guarantee that it will get a value. One alternative could be to calculate the `closestPoint` to the edge of the polygon and use that instead. So even if the point is off of the edge, we use the closest point on the polygon. In fact, I had implemented this method in an old PR that never got merge that you can pull into this PR: https://github.com/UBC-Thunderbots/Software/pull/2813/files#diff-c871a8bcdc98ad35ce2607f7d072014f731fa32dbd28feb62f91d9e2e54ec20b @williamckha do you think think this would meet the requirements of what you're looking for?

Sure, let's do this

Reply:
> the floating point error will probably not be an issue since the `collinear()` method used to check if a point is on a segment uses `almostEqual()` to check for equality. however I agree with Arun as well, crashing the program is probably not a good idea, im not sure if using std::optional is a good idea though.

Actually I'm probably going to add test cases to check for floating point errors